File: app/src/main/java/com/dumuzeyn/mp3player/GlobalSearchOverlayController.java
package com.dumuzeyn.mp3player;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.Button;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import java.util.ArrayList;
import java.util.List;

final class GlobalSearchOverlayController {
    private final MainActivityCore host;
    private final OverlayController overlays;

    GlobalSearchOverlayController(MainActivityCore host, OverlayController overlays) {
        this.host = host;
        this.overlays = overlays;
    }

    void open() {
        FrameLayout shade = host.uiFactory.shade();
        LinearLayout panel = host.uiFactory.panelCard();
        panel.setPadding(host.dp(14), host.dp(12), host.dp(14), host.dp(12));
        panel.addView(header(shade));
        EditText input = new EditText(host);
        input.setHint(host.tr("Songs, artists, albums, genres, playlists",
                "Песни, исполнители, альбомы, жанры, плейлисты"));
        input.setSingleLine(true);
        input.setText(host.navigationState.search);
        input.setTextColor(host.fg);
        input.setHintTextColor(host.muted);
        panel.addView(input, new LinearLayout.LayoutParams(-1, host.dp(54)));
        ScrollView scroll = new ScrollView(host);
        LinearLayout results = new LinearLayout(host);
        results.setOrientation(LinearLayout.VERTICAL);
        scroll.addView(results);
        panel.addView(scroll, new LinearLayout.LayoutParams(-1, 0, 1.0f));
        input.addTextChangedListener(watcher(results, shade));
        shade.addView(panel, host.bottomParams());
        host.overlayHost.addView(shade);
        input.requestFocus();
        submit(results, shade, input.getText().toString());
    }

    private LinearLayout header(FrameLayout shade) {
        LinearLayout row = host.uiFactory.row();
        row.addView(host.uiFactory.text(host.tr("Search", "Поиск"), 22, true),
                new LinearLayout.LayoutParams(0, host.dp(52), 1.0f));
        Button close = host.uiFactory.icon("×");
        close.setContentDescription(host.tr("Close", "Закрыть"));
        close.setOnClickListener(view -> close(shade));
        row.addView(close, host.uiFactory.square(48));
        return row;
    }

    private TextWatcher watcher(LinearLayout results, FrameLayout shade) {
        return new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
            @Override public void afterTextChanged(Editable s) { }
            @Override public void onTextChanged(CharSequence s, int start, int before, int count) {
                submit(results, shade, s == null ? "" : s.toString());
            }
        };
    }

    private void submit(LinearLayout results, FrameLayout shade, String query) {
        host.globalSearchController.search(host.libraryState.tracks,
                host.libraryState.playlists, query, value -> render(results, shade, value));
    }

    private void render(LinearLayout target, FrameLayout shade, GlobalSearchResult result) {
        target.removeAllViews();
        addTracks(target, shade, host.tr("Songs", "Песни"), result.songs);
        addGroups(target, shade, host.tr("Artists", "Исполнители"), result.artists,
                track -> track.artist);
        addGroups(target, shade, host.tr("Albums", "Альбомы"), result.albums,
                track -> track.album);
        addGroups(target, shade, host.tr("Genres", "Жанры"), result.genres,
                track -> track.genre);
        if (!result.playlists.isEmpty()) {
            addHeading(target, host.tr("Playlists", "Плейлисты"));
            for (Playlist playlist : result.playlists) {
                addButton(target, playlist.name, () -> {
                    close(shade);
                    overlays.openPlaylist(playlist);
                });
            }
        }
    }

    private void addTracks(LinearLayout target, FrameLayout shade, String title,
            List<Track> tracks) {
        if (tracks.isEmpty()) {
            return;
        }
        addHeading(target, title);
        for (Track track : tracks) {
            target.addView(host.songsRenderer.songRow(track, true, false,
                    () -> close(shade)));
        }
    }

    private void addGroups(LinearLayout target, FrameLayout shade, String title,
            List<String> values, GroupValue valueProvider) {
        if (values.isEmpty()) {
            return;
        }
        addHeading(target, title);
        for (String value : values) {
            addButton(target, value, () -> {
                close(shade);
                overlays.openGroup(value, groupTracks(value, valueProvider));
            });
        }
    }

    private ArrayList<Track> groupTracks(String value, GroupValue valueProvider) {
        ArrayList<Track> tracks = new ArrayList<>();
        String expected = Track.normalizeSearchText(value);
        for (Track track : host.libraryState.tracks) {
            if (expected.equals(Track.normalizeSearchText(valueProvider.value(track)))) {
                tracks.add(track);
            }
        }
        return tracks;
    }

    private void addHeading(LinearLayout target, String value) {
        target.addView(host.uiFactory.text(value, 15, true),
                new LinearLayout.LayoutParams(-1, host.dp(38)));
    }

    private void addButton(LinearLayout target, String label, Runnable action) {
        Button button = host.uiFactory.button(label);
        button.setOnClickListener(view -> action.run());
        target.addView(button, new LinearLayout.LayoutParams(-1, host.dp(50)));
    }

    private void close(FrameLayout shade) {
        host.globalSearchController.cancel();
        if (shade.getParent() != null) {
            host.overlayHost.removeView(shade);
        }
        host.playerUiController.updateMini();
    }

    private interface GroupValue {
        String value(Track track);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/GlobalSearchOverlayController.kt
package com.dumuzeyn.mp3player

import android.text.Editable
import android.text.TextWatcher
import android.widget.Button
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.ScrollView

internal class GlobalSearchOverlayController(
    private val host: MainActivityCore,
    private val overlays: OverlayController,
) {
    fun open() {
        val shade = host.uiFactory.shade()
        val panel = host.uiFactory.panelCard().apply {
            setPadding(host.dp(14), host.dp(12), host.dp(14), host.dp(12))
            addView(header(shade))
        }
        val input = EditText(host).apply {
            hint = host.tr(
                "Songs, artists, albums, genres, playlists",
                "Песни, исполнители, альбомы, жанры, плейлисты",
            )
            setSingleLine(true)
            setText(host.navigationState.search)
            setTextColor(host.fg)
            setHintTextColor(host.muted)
        }
        panel.addView(input, LinearLayout.LayoutParams(-1, host.dp(54)))
        val results = LinearLayout(host).apply { orientation = LinearLayout.VERTICAL }
        val scroll = ScrollView(host).apply { addView(results) }
        panel.addView(scroll, LinearLayout.LayoutParams(-1, 0, 1f))
        input.addTextChangedListener(watcher(results, shade))
        shade.addView(panel, host.bottomParams())
        host.overlayHost.addView(shade)
        input.requestFocus()
        submit(results, shade, input.text.toString())
    }

    private fun header(shade: FrameLayout): LinearLayout = host.uiFactory.row().apply {
        addView(
            host.uiFactory.text(host.tr("Search", "Поиск"), 22, true),
            LinearLayout.LayoutParams(0, host.dp(52), 1f),
        )
        val close = host.uiFactory.icon("×").apply {
            contentDescription = host.tr("Close", "Закрыть")
            setOnClickListener { close(shade) }
        }
        addView(close, host.uiFactory.square(48))
    }

    private fun watcher(results: LinearLayout, shade: FrameLayout): TextWatcher =
        object : TextWatcher {
            override fun beforeTextChanged(text: CharSequence?, start: Int, count: Int, after: Int) = Unit

            override fun afterTextChanged(text: Editable?) = Unit

            override fun onTextChanged(text: CharSequence?, start: Int, before: Int, count: Int) {
                submit(results, shade, text?.toString().orEmpty())
            }
        }

    private fun submit(results: LinearLayout, shade: FrameLayout, query: String) {
        host.globalSearchController.search(
            host.libraryState.tracks,
            host.libraryState.playlists,
            query,
        ) { result -> render(results, shade, result) }
    }

    private fun render(target: LinearLayout, shade: FrameLayout, result: GlobalSearchResult) {
        target.removeAllViews()
        addTracks(target, shade, host.tr("Songs", "Песни"), result.songs)
        addGroups(target, shade, host.tr("Artists", "Исполнители"), result.artists) { it.artist }
        addGroups(target, shade, host.tr("Albums", "Альбомы"), result.albums) { it.album }
        addGroups(target, shade, host.tr("Genres", "Жанры"), result.genres) { it.genre }
        if (result.playlists.isNotEmpty()) {
            addHeading(target, host.tr("Playlists", "Плейлисты"))
            result.playlists.forEach { playlist ->
                addButton(target, playlist.name) {
                    close(shade)
                    overlays.openPlaylist(playlist)
                }
            }
        }
    }

    private fun addTracks(
        target: LinearLayout,
        shade: FrameLayout,
        title: String,
        tracks: List<Track>,
    ) {
        if (tracks.isEmpty()) return
        addHeading(target, title)
        tracks.forEach { track ->
            target.addView(host.songsRenderer.songRow(track, true, false) { close(shade) })
        }
    }

    private fun addGroups(
        target: LinearLayout,
        shade: FrameLayout,
        title: String,
        values: List<String>,
        valueProvider: GroupValue,
    ) {
        if (values.isEmpty()) return
        addHeading(target, title)
        values.forEach { value ->
            addButton(target, value) {
                close(shade)
                overlays.openGroup(value, groupTracks(value, valueProvider))
            }
        }
    }

    private fun groupTracks(value: String, valueProvider: GroupValue): ArrayList<Track> {
        val expected = Track.normalizeSearchText(value)
        return host.libraryState.tracks.filterTo(ArrayList()) { track ->
            expected == Track.normalizeSearchText(valueProvider.value(track))
        }
    }

    private fun addHeading(target: LinearLayout, value: String) {
        target.addView(
            host.uiFactory.text(value, 15, true),
            LinearLayout.LayoutParams(-1, host.dp(38)),
        )
    }

    private fun addButton(target: LinearLayout, label: String, action: () -> Unit) {
        val button: Button = host.uiFactory.button(label).apply {
            setOnClickListener { action() }
        }
        target.addView(button, LinearLayout.LayoutParams(-1, host.dp(50)))
    }

    private fun close(shade: FrameLayout) {
        host.globalSearchController.cancel()
        if (shade.parent != null) host.overlayHost.removeView(shade)
        host.playerUiController.updateMini()
    }

    private fun interface GroupValue {
        fun value(track: Track): String
    }
}
